### PR TITLE
Implement code2class

### PR DESCRIPTION
**Pull request by SpirinEgor**

### CODE REVIEW

File: configs/__init__.py [hunk 1]
```diff
@@ -62,33 +56,67 @@ def get_code2seq_default_config(dataset_main_folder: str) -> Code2SeqConfig:
...
+    return ModelHyperparameters(
         train_data_path=join(dataset_main_folder, "train"),
         val_data_path=join(dataset_main_folder, "val"),
         test_data_path=join(dataset_main_folder, "test"),
-        encoder=encoder,
-        decoder=decoder,
         n_epochs=5,
         patience=3,
-        batch_size=10,
+        batch_size=25,
```
Comment (maximzubkov) [inline configs/__init__.py:91]: I guess it is better not to change hyperparams even in test config

File: model/code2seq.py [hunk 2]
```diff
@@ -1,40 +1,32 @@
...
 
 
-class Code2Seq(LightningModule):
-    def __init__(self, config: Code2SeqConfig, vocab: Vocabulary, num_workers: int = 0):
-        super().__init__()
+class Code2Seq(BaseCodeModel):
+    def __init__(
+        self, config: Code2SeqConfig, vocab: Vocabulary, num_workers: int,
+    ):
+        super().__init__(config.hyperparams, vocab, num_workers)
```
Comment (maximzubkov) [inline model/code2seq.py:18]: It can be rewritten in a single line

File: model/code2seq.py [hunk 3]
```diff
@@ -85,102 +60,38 @@ def _calculate_loss(self, logits: torch.Tensor, labels: torch.Tensor) -> torch.T
...
         # [seq length; batch size; vocab size]
         logits = self(batch.context, batch.contexts_per_label, batch.labels.shape[0], batch.labels)
         loss = self._calculate_loss(logits, batch.labels)
-
+        log = {"train/loss": loss}
         with torch.no_grad():
-            subtoken_statistic = SubtokenStatistic.calculate_statistic(
-                batch.labels, logits.argmax(-1), [self.vocab.label_to_id[t] for t in [SOS, EOS, PAD, UNK]]
-            )
+            statistic = SubtokenStatistic().calculate_statistic(batch.labels, logits.argmax(-1))
```
Comment (maximzubkov) [inline model/code2seq.py:77]: Maybe it will be more correct  to change the function `calculate_statistic` signature from `labels, logits` to ones similar to `cross_entopy`, namely: `logits, labels`

File: model/code2seq.py [hunk 4]
```diff
@@ -1,40 +1,32 @@
...
+            config.decoder_config.decoder_size,
+            len(self.vocab.token_to_id),
+            self.vocab.token_to_id[PAD],
+            len(self.vocab.type_to_id),
+            self.vocab.type_to_id[PAD],
         )
         self.decoder = PathDecoder(
-            decoder_config, len(vocab.label_to_id), vocab.label_to_id[SOS], vocab.label_to_id[PAD]
+            config.decoder_config, len(self.vocab.label_to_id), self.vocab.label_to_id[SOS], self.vocab.label_to_id[PAD]
         )
```
Comment (maximzubkov) [inline model/code2seq.py:30]: Maybe it will be better to use variables `encoder_config, decoder_config`, as it was before?

File: model/code2seq.py (hunk 3 above)
Comment (SpirinEgor) [reply model/code2seq.py:77]: It doesn't correspond to Code2Class model. We can do it later, in another pr.

File: model/__init__.py [hunk 5]
```diff
@@ -1 +1,3 @@
+from .base_code_model import BaseCodeModel
```
Comment (SpirinEgor) [inline model/__init__.py:1]: `BaseCodeModel` used as an abstract class, so it shouldn't be visible for users. We can remove it from here or use `__all__`

File: model/modules/__init__.py [hunk 6]
```diff
@@ -1,3 +1,4 @@
+from .attention import LuongAttention
```
Comment (SpirinEgor) [inline model/modules/__init__.py:1]: What about LocalAttention?
It seems that we should import all attention layers or not import them at all.

File: model/modules/path_classifier.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        """Classify given paths
+
+        :param encoded_paths: [n paths; classifier size]
+        :param contexts_per_label: [n1, n2, ..., nk] sum = n paths
+        :return:
+        """
+        # [batch size; max context size; classifier input size], [batch size; max context size]
+        batched_context, attention_mask = cut_encoded_contexts(encoded_paths, contexts_per_label, self._negative_value)
+
+        # [batch size; classifier input size]
```
Comment (SpirinEgor) [inline model/modules/path_classifier.py:43]: `# [batch size; max context size; 1]`

File: model/modules/path_classifier.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+        attn_weights = self.attention(batched_context, attention_mask)
+
+        # [batch size; classifier input size]
+        context = torch.sum(batched_context * attn_weights, dim=1)
```
Comment (SpirinEgor) [inline model/modules/path_classifier.py:47]: Hmm, how it works? If attention is 3-dim tensor then we should use `bmm`. Also, this will sum vectors.
```
context = torch.bmm(attn_weights.transpose(1, 2), batched_context).squeeze(1)
```
